### PR TITLE
[FLINK-24918][Runtime/State Backends]Support to specify the data dir for state benchmark

### PR DESCRIPTION
**Pull request by Aitozi**
Rely on the Flink [PR](https://github.com/apache/flink/pull/17846)

### CODE REVIEW

**Comment by Aitozi** [comment]
This build error due to it reply on the new api in [PR](https://github.com/apache/flink/pull/17846)

File: README.md
Comment (pnowojski) [inline README.md:31]: nitty nit: In line 24 replace:
> There're mainly two way

with

> There're three ways

?

File: README.md
Comment (pnowojski) [inline README.md:49]: nit: (here and in other places) `params` -> `parameters`

File: src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java
Comment (pnowojski) [inline src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java:61]: Will this be behaving in the same/similar way as it used to be? What if on the given machine there is no `/tmp` directory or the JVM process doesn't have an access to it? I think the old `null` value was more resilient?

**Review by pnowojski — COMMENTED** [review]
+1 for the change. I've left a couple of comments.

File: src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java
Comment (Aitozi) [reply src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java]: It's indeed have different behavior, because it's depends on the `java.io.tmpdir` property. 
The `@param` do not accept `null`, so I use `not_set` as default value . 

```
# Warmup: 10 iterations, 1 s each
# Measurement: 10 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: org.apache.flink.state.benchmark.ValueStateBenchmark.valueAdd
# Parameters: (backendType = HEAP, stateDataDirPath = not_set)
```

**Comment by Aitozi** [comment]
@pnowojski  Thanks for your review, I have addressed your comments.

File: README.md
Comment (pnowojski) [inline README.md:51]: Missed `params`

File: README.md
Comment (pnowojski) [inline README.md:54]: Missed `params`

File: src/main/java/org/apache/flink/state/benchmark/ListStateBenchmark.java
Comment (pnowojski) [inline src/main/java/org/apache/flink/state/benchmark/ListStateBenchmark.java:29]: Why is this needed?

File: src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java
Comment (pnowojski) [inline src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java:60]: Won't this in combination with https://github.com/apache/flink-benchmarks/blob/master/save_jmh_result.py#L72:L76 change the names of the uploaded benchmarks to the [code speed](http://codespeed.dak8s.net:8000/changes/)?

Do we really want this to be reported like this? Won't this clog the results and the WebUI in particular?
Secondly, even if we want this, we would have to manually rename all of the benchmarks in the database to preserve continuity of the results in the graphs.

Maybe it's actually better idea after all, to not use jmh's `@Param` for this? Maybe simply use an environmental variable?

File: src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java
Comment (pnowojski) [inline src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java:23]: Are you using the same formatting settings as for the main Flink repository? 

https://nightlies.apache.org/flink/flink-docs-master/docs/flinkdev/ide_setup/#code-formatting

**Review by pnowojski — COMMENTED** [review]
Thanks for the updates. I've left a couple of minor comments, and one more issue that I've just realised about.

File: src/main/java/org/apache/flink/state/benchmark/ListStateBenchmark.java
Comment (Aitozi) [reply src/main/java/org/apache/flink/state/benchmark/ListStateBenchmark.java:29]: removed

File: src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java
Comment (Aitozi) [reply src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java:23]: updated

File: src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java
Comment (Aitozi) [reply src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java:61]: I didn't notice it. Now I prefer to not use `@Param` for this . Instead of using environmental variable, Can we add a benchmark conf as I mentioned in [FLINK-24967](https://issues.apache.org/jira/projects/FLINK/issues/FLINK-24967)

Comment (pnowojski) [reply src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java:61]: I think either of those options would work, depends on what we want to achieve? I think using `@Params` makes sense if we really want to present the results from different configurations side by side, especially in the WebUI or in the continuous regression testing scripts. However if there are many many different constants that you would like to change, it probably doesn't make sense to keep them all separately as `@Params`, as it would simply explode the number of tests (jmh is testing all combinations of params). 

Having said all of that, a dedicated config file might be a not a bad solution. For now there can be a single configuration file that anyone can overwrite as he chooses. If we ever decide to present the results from different configs side by side, we can do this via `@Param` `preset`, where `preset` would be a name of one of those config files.

Comment (Aitozi) [reply src/main/java/org/apache/flink/state/benchmark/StateBenchmarkBase.java:61]: ok, I will provide a single configuration implementation .

**Comment by Aitozi** [comment]
@pnowojski Sorry for the late response, I have done the work to add support for a dedicated config file. Please take a look when you are free. 

I have check the ci failure log , It seems unrelated to this Pr.

```
 Could not find artifact org.apache.flink:flink-examples-batch_2.12:jar:tests:1.15-SNAPSHOT in sonatype-snapshots (https://oss.sonatype.org/content/repositories/snapshots/) -> [Help 1]
```

**Review by pnowojski — APPROVED** [review]
LGTM % two nits

**Comment by Aitozi** [comment]
Thanks for your comments @pnowojski , I have updated the PR accordingly

**Review by pnowojski — APPROVED** [review]
Thanks for the update. LGTM,. Can you squash the 2nd, 3rd and 4th commits together? Do you have committer writes and can you merge this PR on your own, or should I do it?

**Comment by Aitozi** [comment]
@pnowojski I squashed the commit, I do not have committer writers, please help to merge it 😄

**Comment by pnowojski** [comment]
Thanks @Aitozi. I will do it, but could you take a look at the travis failure? It looks probably unrelated to your change, but it blocks merging of this PR.

**Comment by Aitozi** [comment]
OK, I will try to look into it tonight.

**Comment by Aitozi** [comment]
It should be caused by [[FLINK-25039][build] Disable shading of test jar by default]( https://github.com/apache/flink/pull/17895 ). After that, the `flink-examples-batch_2.12` will not generate an empty test-jar. 

Since the test-jar is empty, so we do not depend on it actually. I think it can be removed safely. Let's see will the travis turn green :)

**Comment by pnowojski** [comment]
Thanks! LGTM. Can you just rename the last commit by prefixing it with `[FLINK-25039]`?

**Comment by Aitozi** [comment]
done @pnowojski

**Comment by pnowojski** [comment]
Thanks :) merging

**Pull request merged**
